cmd/juju/user/login.go

```diff
@@ -132,142 +134,86 @@ func (c *loginCommand) Run(ctx *cmd.Context) error {
-		}
-	}
-	if !names.IsValidUserName(user) {
-		return errors.NotValidf("user name %q", user)
+	if strings.Contains(c.controllerName, ":") {
+		return errors.Errorf("cannot use %q as a controller name - use -c flag to choose a different one", c.controllerName)
 	}
-	userTag := names.NewUserTag(user)
 
-	// Make sure that the client is not already logged in,
-	// or if it is, that it is logged in as the specified
-	// user.
-	if accountDetails != nil && accountDetails.User != userTag.Id() {
-		return errors.New(`already logged in
-
-Run "juju logout" first before attempting to log in as a different user.`)
-	}
-
-	// Log in without specifying a password in the account details. This
-	// will trigger macaroon-based authentication, which will prompt the
-	// user for their password.
-	accountDetails = &jujuclient.AccountDetails{
-		User: userTag.Id(),
-	}
-	params, err := c.NewAPIConnectionParams(store, controllerName, "", accountDetails)
-	if err != nil {
+	// Find out details on the specified controller.
+	controllerDetails, err := store.ControllerByName(c.controllerName)
+	if err != nil && !errors.IsNotFound(err) {
 		return errors.Trace(err)
 	}
-	conn, err := newAPIConnection(params)
-	if err != nil {
-		return errors.Annotate(err, "creating API connection")
-	}
-	defer conn.Close()
 
-	accountDetails.LastKnownAccess = conn.ControllerAccess()
-	if err := store.UpdateAccount(controllerName, *accountDetails); err != nil {
-		return errors.Annotate(err, "failed to record temporary credential")
+	// Find out details of the controller domain if it's specified.
+	var conn api.Connection
+	var publicControllerDetails *jujuclient.ControllerDetails
+	var accountDetails *jujuclient.AccountDetails
+	// Fetch current details for the specified controller name so we
+	// can tell if the logged in user has changed.
+	oldAccountDetails, err := store.AccountDetails(c.controllerName)
+	if err != nil && !errors.IsNotFound(err) {
+		return errors.Trace(err)
 	}
-	ctx.Infof("You are now logged in to %q as %q.", controllerName, userTag.Id())
-	return nil
-}
-
-func (c *loginCommand) controllerLogin(ctx *cmd.Context, store jujuclient.ClientStore) error {
-	controllerHost := c.userOrDomain
-	if !strings.Contains(controllerHost, ".") {
-		// TODO(rog) provide a way of avoiding this external network access.
-		controllerHost1, err := c.getKnownControllerDomain(controllerHost)
-		if errors.IsNotFound(err) {
-			return errors.Errorf("%q is not a known public controller", controllerHost)
+	if c.domain != "" {
+		conn, publicControllerDetails, accountDetails, err = c.publicControllerLogin(ctx, c.domain, oldAccountDetails)
+		if err != nil {
+			return errors.Annotatef(err, "cannot log into %q", c.domain)
 		}
+	} else if controllerDetails == nil {
+		// No controller found and no domain specified - we
+		// have no idea where we should be logging in.
+		return errors.Errorf("domain not specified and controller does not exist")
```

Comment:
Unfortunately, "host name" isn't accurate either, as it can refer to one of the publicly defined controller aliases. "no current controller registered" isn't quite right either, as when the -c flag is specified we might not be talking about the current controller. I'll try to think of something else.